### PR TITLE
deprecate Task.didWork

### LINKED ISSUE #23460: Depracete Task.didWork
The definition of when a task "did work" is a bit confusing, and is not very useful. Let's deprecate `Task.didWork`. See reasoning in https://github.com/gradle/gradle/issues/21856.

---
cc: @gradle/execution

### PR DESCRIPTION
**Pull request by tschulte**
This commit only deprecates Task.didWork and TaskState.didWork. Internally the current behavior is unchanged (#22120 will change that).

Fixes #23460

### Context
The definition of when a task "did work" is a bit confusing, and is not very useful. Let's deprecate Task.didWork. See reasoning in https://github.com/gradle/gradle/issues/21856.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 1]
```diff
@@ -84,6 +84,23 @@ Some builds may have worked around this by also putting the additional attribute
 
 The embedded Kotlin has been updated to link:https://github.com/JetBrains/kotlin/releases/tag/v1.8.21[Kotlin 1.8.21].
 
+=== Deprecations
+
+[[deprecated_task_did_work]]
+==== Deprecated Task.didWork
+
+The following methods on `Task` are deprecated
```
Comment (lkasso) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:92]:
```suggestion
The following methods on `Task` are deprecated:
```

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 2]
```diff
@@ -84,6 +84,23 @@ Some builds may have worked around this by also putting the additional attribute
...
+
+- link:{javadocPath}/org/gradle/api/Task.html#setDidWork-boolean-[setDidWork]
+- link:{javadocPath}/org/gradle/api/Task.html#getDidWork--[getDidWork]
+
+For some tasks, `Task.getDidWork` returns false even if the task did execute but the output
```
Comment (lkasso) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:97]:
```suggestion
For some tasks, `Task.getDidWork` returns false even if the task executed and the output(s)
```

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 3]
```diff
@@ -84,6 +84,23 @@ Some builds may have worked around this by also putting the additional attribute
...
+did not change (aka "up-to-date after-the-fact"). This is unexpected, because although the
```
Comment (lkasso) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:98]:
```suggestion
did not change (aka "up-to-date after-the-fact"). Although the
```

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 4]
```diff
@@ -84,6 +84,23 @@ Some builds may have worked around this by also putting the additional attribute
...
+task was executed and took time, the output of the task is `UP-TO-DATE`. This will change in
```
Comment (lkasso) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:99]:
```suggestion
task executed and took time, the outcome label of the task is `UP-TO-DATE`. 
This behavior is unexpected.
```

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 5]
```diff
@@ -84,6 +84,23 @@ Some builds may have worked around this by also putting the additional attribute
...
+the future, and as a step in that direction, `Task.didWork` was deprecated. You can check if
```
Comment (lkasso) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:100]:
```suggestion
As part of ongoing efforts to fix this behavior, `Task.didWork` was deprecated. 
```

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 6]
```diff
@@ -84,6 +84,23 @@ Some builds may have worked around this by also putting the additional attribute
...
+the task was "up-to-date up front" by using
```
Comment (lkasso) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:101]:
```suggestion
You can check if a task was "up-to-date up front" by using
```

**Review by lkasso — CHANGES_REQUESTED** [review]
Some minor doc cleanup (feel free to make further corrections if needed).

**Comment by lptr** [comment]
Hi @tschulte, thank you for the effort. Unfortunately we were confusing in putting the "onboarding" label on this issue; it's not an area we are focusing on right now. I will close this PR for now, and will come back to it once we pick up the work on dropping `Task.didWork`. Sorry about misleading you. :(

If you'd like to contribute to Gradle, we now have a good list of potential first issues here: https://github.com/gradle/gradle/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22

**Pull request closed without merging**